pkg/provisioners/fanout/fanout_handler.go

```diff
@@ -108,10 +126,35 @@ func (f *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 // requests return successfully, then return nil. Else, return an error.
 func (f *Handler) dispatch(msg *provisioners.Message) error {
 	errorCh := make(chan error, len(f.config.Subscriptions))
-	for _, sub := range f.config.Subscriptions {
-		go func(s eventingduck.SubscriberSpec) {
-			errorCh <- f.makeFanoutRequest(*msg, s)
-		}(sub)
+
+	if f.config.Router != nil {
+		routes, err := f.dispatcher.ComputeRoutes(msg, f.config.Router.RouterURI, provisioners.DispatchDefaults{})
+		if err != nil {
+			f.logger.Error("Compute routes had an error", zap.Error(err))
+			return err
+		}
+		subsCount := len(f.config.Subscriptions)
+
+		for _, name := range routes {
+			i, ok := f.subscribers[name]
+			if !ok {
+				errorCh <- fmt.Errorf("Invalid route %v", name)
+			} else {
+				go func(s eventingduck.SubscriberSpec) {
+					errorCh <- f.makeFanoutRequest(*msg, s)
+				}(f.config.Subscriptions[i])
+			}
+		}
+
+		for i := subsCount - len(routes); i >= 0; i-- {
```

Comment:
IMO, this might be simpler to have routes be a set, and then iterate over subscribers and compare the name with the set. If no match, you can send back a nil. If there is a match, you can pop from the set and return the result of the request.

If you want to log any names that don't match Subscriptions as errors, you can then log them directly, and the accounting is a bit simpler to track, I think.

Reply:
ideally we want to have another method to handle the router case so that we don't have to iterate over all subscriptions. I mean this could also be done in `dispatch`, but that'd cleaner.